src/charm.py

```diff
@@ -1,303 +1,311 @@
+                                    charm.
+                                    NOTE: This will only be used if initial regular apply() fails.
+        """
+        self.unit.status = MaintenanceStatus("Creating K8S resources")
+        try:
+            self.k8s_resource_handler.apply()
+        except ApiError as error:
+            if self._check_and_report_k8s_conflict(error) and force_conflicts:
+                # conflict detected when applying K8S resources
+                # re-apply K8S resources with forced conflict resolution
+                self.unit.status = MaintenanceStatus("Force applying K8S resources")
+                self.logger.warning("Apply K8S resources with forced changes against conflicts")
+                self.k8s_resource_handler.apply(force=force_conflicts)
+            else:
+                raise GenericCharmRuntimeError("K8S resources creation failed") from error
+        try:
+            self.crd_resource_handler.apply()
+        except ApiError as error:
+            if self._check_and_report_k8s_conflict(error) and force_conflicts:
+                # conflict detected when applying CRD resources
+                # re-apply CRD resources with forced conflict resolution
+                self.unit.status = MaintenanceStatus("Force applying CRD resources")
+                self.logger.warning("Apply CRD resources with forced changes against conflicts")
+                self.crd_resource_handler.apply(force=force_conflicts)
+            else:
+                raise GenericCharmRuntimeError("CRD resources creation failed") from error
+        self.model.unit.status = MaintenanceStatus("K8S resources created")
+
+    def _gen_certs_if_missing(self):
+        """Generate certificates if they don't already exist in _stored."""
+        self.logger.info("Generating certificates if missing.")
+        cert_attributes = ["cert", "ca", "key"]
+        # Generate new certs if any cert attribute is missing
+        for cert_attribute in cert_attributes:
+            try:
+                getattr(self._stored, cert_attribute)
+            except AttributeError:
+                self._gen_certs()
+                break
+        self.logger.info("Certificates already exist.")
+
+    def _gen_certs(self):
+        """Refresh the certificates, overwriting them if they already existed."""
+        self.logger.info("Generating certificates..")
+        certs = gen_certs(model=self._namespace, service_name=f"{self._service_name}")
+        for k, v in certs.items():
+            setattr(self._stored, k, v)
+
+    def _upload_certs_to_container(self):
+        """Upload generated certs to container."""
+        try:
+            self._check_container_connection(self.container)
+        except ErrorWithStatus as error:
+            self.model.unit.status = error.status
+            return
+
+        self.container.push(
+            Path(f"{CONTAINER_CERTS_DEST}/key.pem"), self._cert_key, make_dirs=True
```

Comment:
I think it should be different

```suggestion
CONTAINER_CERTS_DEST = Path(...)
CONTAINER_CERTS_DEST / "key.pem", self._cert_key, make_dirs=True
```